core/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -114,28 +113,21 @@ public QueryShardContext(QueryShardContext source) {
         this.types = source.getTypes();
     }
 
-    public void parseFieldMatcher(ParseFieldMatcher parseFieldMatcher) {
-        this.parseContext.parseFieldMatcher(parseFieldMatcher);
-    }
-
+    /**
+     * @return the parseFieldMatcher configured in the index settings
+     */
     public ParseFieldMatcher parseFieldMatcher() {
-        return parseContext.parseFieldMatcher();
+        return indexSettings.getParseFieldMatcher();
     }
 
     public void reset() {
```

Comment:
I'll look into it, but maybe to keep this change from growing too large I would tackle that removal as a follow up.